### PR TITLE
NIFI-13868: add `pg-delete` command to Nifi Toolkit

### PR DESCRIPTION
**Pull request by grishick**
# Summary

[NIFI-13868](https://issues.apache.org/jira/browse/NIFI-13868)

Adding `pg-delete` command to Nifi Toolkit. The command works the same way as deleting a process group via UI. If there are messages in PG queue, deletion will fail. I ported this change from [Anetac's Nifi fork](https://github.com/Eng-Anetac/nifi) where I made it a while ago. We have been using this feature in production at Anetac for months now.

PR with the same feature for `main` branch: https://github.com/apache/nifi/pull/9388
The change for support/nifi-1.x branch is larger than for main, because the support branch does not have changes for NIFI-12898

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `support/nifi-1.x` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-toolkit/nifi-toolkit-cli/src/main/java/org/apache/nifi/toolkit/cli/impl/client/nifi/ProcessGroupClient.java [hunk 1]
```diff
@@ -75,4 +75,6 @@ FlowEntity copySnippet(String processGroupId, CopySnippetRequestEntity copySnipp
         throws NiFiClientException, IOException;
 
     FlowComparisonEntity getLocalModifications(String processGroupId) throws NiFiClientException, IOException;
+
+    void deleteProcessGroup(ProcessGroupEntity entity) throws NiFiClientException, IOException;
```
Comment (exceptionfactory) [inline nifi-toolkit/nifi-toolkit-cli/src/main/java/org/apache/nifi/toolkit/cli/impl/client/nifi/ProcessGroupClient.java:79]: As other delete commands return the deleted entity, that looks like a good pattern to follow in this scenario.

File: nifi-toolkit/nifi-toolkit-cli/src/main/java/org/apache/nifi/toolkit/cli/impl/command/nifi/pg/PGDelete.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+        @Override
+        public VoidResult doExecute(final NiFiClient client, final Properties properties)
+                throws NiFiClientException, IOException, MissingOptionException {
+
+            final String pgId = getRequiredArg(properties, CommandOption.PG_ID);
+
+            final ProcessGroupClient pgClient = client.getProcessGroupClient();
+            final ProcessGroupEntity pgEntity = pgClient.getProcessGroup(pgId);
+            if(pgEntity == null) {
```
Comment (exceptionfactory) [inline nifi-toolkit/nifi-toolkit-cli/src/main/java/org/apache/nifi/toolkit/cli/impl/command/nifi/pg/PGDelete.java:60]: Spacing:
```suggestion
            if (pgEntity == null) {
```

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the contribution @grishick. The general implementation looks good, just one functional note on returning the deleted entity instead of void.

Also, please submit new features to the main branch first. Once that is approved and merged, then submitting a separate pull request for the support branch would be helpful if it is desirable to have this functionality on the support branch.

File: nifi-toolkit/nifi-toolkit-cli/src/main/java/org/apache/nifi/toolkit/cli/impl/client/nifi/ProcessGroupClient.java (hunk 1 above)
Comment (grishick) [reply nifi-toolkit/nifi-toolkit-cli/src/main/java/org/apache/nifi/toolkit/cli/impl/client/nifi/ProcessGroupClient.java]: done

**Comment by grishick** [comment]
> Thanks for the contribution @grishick. The general implementation looks good, just one functional note on returning the deleted entity instead of void.
> 
> Also, please submit new features to the main branch first. Once that is approved and merged, then submitting a separate pull request for the support branch would be helpful if it is desirable to have this functionality on the support branch.

Got it. Here is the PR for main branch: https://github.com/apache/nifi/pull/9388

**Review by pvillard31 — COMMENTED** [review]
Thanks for the improvement. I'd recommend adding more details in the helper text for this command. For this command to be successful, three actions would need to be taken before:
- stop the processors in the process group (`pg-stop`)
- disable the controller services in the process group (`pg-disable-services`)
- empty all queues in the process group (we don't have a CLI command for that now, would certainly be a nice addition)

I didn't test the current PR but would be good to confirm that everything works fine especially when having nested process groups with controller services at multiple levels, etc.

**Comment by grishick** [comment]
Sounds good. I'll try to make the changes asap.

**Review by exceptionfactory — APPROVED** [review]
Thanks @grishick, I corrected formatting on `PGDelete` as in the PR for the main branch. The rest looks good. +1 merging

**Pull request merged**